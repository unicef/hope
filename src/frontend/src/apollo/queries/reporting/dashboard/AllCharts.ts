import { gql } from '@apollo/client';

export const Allcharts = gql`
  query AllCharts(
    $businessAreaSlug: String!
    $year: Int!
    $program: String
    $administrativeArea: String
  ) {
    chartProgrammesBySector(
      businessAreaSlug: $businessAreaSlug
      year: $year
      program: $program
      administrativeArea: $administrativeArea
    ) {
      labels
      datasets {
        label
        data
      }
    }
    chartPaymentVerification(
      businessAreaSlug: $businessAreaSlug
      year: $year
      program: $program
      administrativeArea: $administrativeArea
    ) {
      datasets {
        label
        data
      }
      labels
      households
      averageSampleSize
    }
    chartPaymentVerificationForPeople(
      businessAreaSlug: $businessAreaSlug
      year: $year
      program: $program
      administrativeArea: $administrativeArea
    ) {
      datasets {
        label
        data
      }
      labels
      households
      averageSampleSize
    }
    chartVolumeByDeliveryMechanism(
      businessAreaSlug: $businessAreaSlug
      year: $year
      program: $program
      administrativeArea: $administrativeArea
    ) {
      datasets {
        data
      }
      labels
    }
    chartPayment(
      businessAreaSlug: $businessAreaSlug
      year: $year
      program: $program
      administrativeArea: $administrativeArea
    ) {
      datasets {
        data
      }
      labels
    }
    chartGrievances(
      businessAreaSlug: $businessAreaSlug
      year: $year
      administrativeArea: $administrativeArea
    ) {
      datasets {
        data
      }
      labels
      totalNumberOfGrievances
      totalNumberOfFeedback
      totalNumberOfOpenSensitive
    }
    sectionHouseholdsReached(
      businessAreaSlug: $businessAreaSlug
      year: $year
      program: $program
      administrativeArea: $administrativeArea
    ) {
      total
    }
    sectionIndividualsReached(
      businessAreaSlug: $businessAreaSlug
      year: $year
      program: $program
      administrativeArea: $administrativeArea
    ) {
      total
    }
    sectionPeopleReached(
      businessAreaSlug: $businessAreaSlug
      year: $year
      program: $program
      administrativeArea: $administrativeArea
    ) {
      total
    }
    sectionChildReached(
      businessAreaSlug: $businessAreaSlug
      year: $year
      program: $program
      administrativeArea: $administrativeArea
    ) {
      total
    }

    chartIndividualsReachedByAgeAndGender(
      businessAreaSlug: $businessAreaSlug
      year: $year
      program: $program
      administrativeArea: $administrativeArea
    ) {
      datasets {
        data
      }
      labels
    }
    chartIndividualsWithDisabilityReachedByAge(
      businessAreaSlug: $businessAreaSlug
      year: $year
      program: $program
      administrativeArea: $administrativeArea
    ) {
      datasets {
        data
        label
      }
      labels
    }
    chartPeopleReachedByAgeAndGender(
      businessAreaSlug: $businessAreaSlug
      year: $year
      program: $program
      administrativeArea: $administrativeArea
    ) {
      datasets {
        data
      }
      labels
    }
    chartPeopleWithDisabilityReachedByAge(
      businessAreaSlug: $businessAreaSlug
      year: $year
      program: $program
      administrativeArea: $administrativeArea
    ) {
      datasets {
        data
        label
      }
      labels
    }
    sectionTotalTransferred(
      businessAreaSlug: $businessAreaSlug
      year: $year
      program: $program
      administrativeArea: $administrativeArea
    ) {
      total
    }
    chartTotalTransferredByMonth(
      businessAreaSlug: $businessAreaSlug
      year: $year
      program: $program
      administrativeArea: $administrativeArea
    ) {
      datasets {
        data
        label
      }
      labels
    }
  }
`;
