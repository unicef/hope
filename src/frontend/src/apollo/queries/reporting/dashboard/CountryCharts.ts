import { gql } from '@apollo/client';

export const CountryCharts = gql`
  query CountryCharts(
    $businessAreaSlug: String!
    $year: Int!
    $program: String
    $administrativeArea: String
    $order: String
    $orderBy: String
  ) {
    tableTotalCashTransferredByAdministrativeArea(
      businessAreaSlug: $businessAreaSlug
      year: $year
      program: $program
      administrativeArea: $administrativeArea
      order: $order
      orderBy: $orderBy
    ) {
      data {
        id
        admin2
        totalCashTransferred
        totalHouseholds
      }
    }
    tableTotalCashTransferredByAdministrativeAreaForPeople(
      businessAreaSlug: $businessAreaSlug
      year: $year
      program: $program
      administrativeArea: $administrativeArea
      order: $order
      orderBy: $orderBy
    ) {
      data {
        id
        admin2
        totalCashTransferred
        totalPeople
      }
    }
  }
`;
