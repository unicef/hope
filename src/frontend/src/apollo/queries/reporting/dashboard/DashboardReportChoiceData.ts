import { gql } from '@apollo/client';

export const DashboardReportChoiceData = gql`
  query DashboardReportChoiceData($businessArea: String!) {
    dashboardReportTypesChoices(businessAreaSlug: $businessArea) {
      name
      value
    }
  }
`;
