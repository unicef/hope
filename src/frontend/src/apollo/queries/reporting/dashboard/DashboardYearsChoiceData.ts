import { gql } from '@apollo/client';

export const DashboardYearsChoiceData = gql`
  query DashboardYearsChoiceData($businessArea: String!) {
    dashboardYearsChoices(businessAreaSlug: $businessArea)
  }
`;
