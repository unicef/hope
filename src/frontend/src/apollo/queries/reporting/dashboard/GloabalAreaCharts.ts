import { gql } from '@apollo/client';

export const GlobalAreaCharts = gql`
  query GlobalAreaCharts($year: Int!) {
    chartTotalTransferredCashByCountry(year: $year) {
      datasets {
        data
        label
      }
      labels
    }
  }
`;
