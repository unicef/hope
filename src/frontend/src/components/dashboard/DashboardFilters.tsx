import { Grid } from '@mui/material';
import * as React from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import { useBaseUrl } from '@hooks/useBaseUrl';
import { AdminAreaAutocomplete } from '@shared/autocompletes/AdminAreaAutocomplete';
import { createHandleApplyFilterChange } from '@utils/utils';
import { FiltersSection } from '@core/FiltersSection';
import { ProgramAutocomplete } from '@shared/autocompletes/ProgramAutocomplete';

interface DashboardFiltersProps {
  filter;
  setFilter;
  initialFilter;
  appliedFilter;
  setAppliedFilter;
}

export const DashboardFilters = ({
  filter,
  setFilter,
  initialFilter,
  appliedFilter,
  setAppliedFilter,
}: DashboardFiltersProps): React.ReactElement => {
  const navigate = useNavigate();
  const location = useLocation();
  const { isAllPrograms } = useBaseUrl();

  const { applyFilterChanges, clearFilter } = createHandleApplyFilterChange(
    initialFilter,
    navigate,
    location,
    filter,
    setFilter,
    appliedFilter,
    setAppliedFilter,
  );
  const handleApplyFilter = (): void => {
    applyFilterChanges();
  };

  const handleClearFilter = (): void => {
    clearFilter();
  };

  return (
    <FiltersSection
      clearHandler={handleClearFilter}
      applyHandler={handleApplyFilter}
    >
      <Grid container alignItems="flex-end" spacing={3}>
        {isAllPrograms && (
          <Grid item xs={5}>
            <ProgramAutocomplete
              filter={filter}
              name="program"
              value={filter.program}
              setFilter={setFilter}
              initialFilter={initialFilter}
              appliedFilter={appliedFilter}
              setAppliedFilter={setAppliedFilter}
            />
          </Grid>
        )}
        <Grid item xs={3}>
          <AdminAreaAutocomplete
            name="administrativeArea"
            level={2}
            value={filter.administrativeArea}
            filter={filter}
            setFilter={setFilter}
            initialFilter={initialFilter}
            appliedFilter={appliedFilter}
            setAppliedFilter={setAppliedFilter}
            dataCy="filter-administrative-area"
          />
        </Grid>
      </Grid>
    </FiltersSection>
  );
};
