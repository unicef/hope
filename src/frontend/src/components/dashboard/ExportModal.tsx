import {
  Box,
  Button,
  Checkbox,
  DialogContent,
  DialogTitle,
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableRow,
  Typography,
} from '@mui/material';
import * as React from 'react';
import { useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Dialog } from '@containers/dialogs/Dialog';
import { DialogActions } from '@containers/dialogs/DialogActions';
import { DialogContainer } from '@containers/dialogs/DialogContainer';
import { DialogFooter } from '@containers/dialogs/DialogFooter';
import { DialogTitleWrapper } from '@containers/dialogs/DialogTitleWrapper';
import { useSnackbar } from '@hooks/useSnackBar';
import {
  useCreateDashboardReportMutation,
  useDashboardReportChoiceDataQuery,
} from '@generated/graphql';
import { LoadingComponent } from '@core/LoadingComponent';
import { useBaseUrl } from '@hooks/useBaseUrl';

export function ExportModal({ filter, year }): React.ReactElement {
  const [dialogOpen, setDialogOpen] = useState(false);
  const [selected, setSelected] = useState([]);
  const { businessArea, programId, isAllPrograms } = useBaseUrl();
  const { showMessage } = useSnackbar();
  const { t } = useTranslation();
  const numSelected = selected.length;
  const isSelected = (id: string): boolean => selected.includes(id);

  const { data: choicesData, loading: choicesLoading } =
    useDashboardReportChoiceDataQuery({ variables: { businessArea } });
  const [mutate] = useCreateDashboardReportMutation();

  useEffect(() => {
    setSelected([]);
  }, [businessArea]);

  if (choicesLoading) return <LoadingComponent />;
  if (!choicesData) return null;

  const data = choicesData.dashboardReportTypesChoices.map((choice) => ({
    id: choice.value,
    name: choice.name,
  }));

  const rowCount = data.length;

  const onSelectAllClick = (event, rows): void => {
    if (event.target.checked) {
      const newSelecteds = rows.map((row) => row.id);
      setSelected(newSelecteds);
      return;
    }
    setSelected([]);
  };
  const onCheckboxClick = (id: string): void => {
    const selectedIndex = selected.indexOf(id);
    const newSelected = [...selected];
    if (selectedIndex !== -1) {
      newSelected.splice(selectedIndex, 1);
    } else {
      newSelected.push(id);
    }
    setSelected(newSelected);
  };

  const renderRows = (): Array<React.ReactElement> =>
    data.map((el) => {
      const isItemSelected = isSelected(el.id);
      return (
        <TableRow onClick={() => onCheckboxClick(el.id)} key={el.id}>
          <TableCell align="left" padding="checkbox">
            <Checkbox
              color="primary"
              onClick={() => onCheckboxClick(el.id)}
              checked={isItemSelected}
              inputProps={{ 'aria-labelledby': el.id }}
            />
          </TableCell>
          <TableCell align="left">{el.name}</TableCell>
        </TableRow>
      );
    });

  const submitFormHandler = (): void => {
    mutate({
      variables: {
        reportData: {
          businessAreaSlug: businessArea,
          reportTypes: selected,
          year: parseInt(year, 10),
          adminArea: filter.administrativeArea?.node?.id,
          program: isAllPrograms ? null : programId,
        },
      },
    })
      .then((response) => {
        if (!response.errors && response.data.createDashboardReport.success) {
          showMessage(t('Report was created.'));
        } else {
          showMessage(t('Report create action failed.'));
        }
        setSelected([]);
        setDialogOpen(false);
      })
      .catch((error) => {
        console.error(error);
      });
  };

  return (
    <>
      <Button
        color="primary"
        variant="contained"
        onClick={() => setDialogOpen(true)}
        data-cy="button-ed-plan"
      >
        Export
      </Button>
      <Dialog
        open={dialogOpen}
        onClose={() => setDialogOpen(false)}
        scroll="paper"
        aria-labelledby="form-dialog-title"
        maxWidth="md"
      >
        <DialogTitleWrapper>
          <DialogTitle>{t('Export Data')}</DialogTitle>
        </DialogTitleWrapper>
        <DialogContent>
          <DialogContainer>
            <Box mb={6}>
              <Typography variant="body2">
                {t(
                  'The filters applied on the dashboard will be used for the reports.',
                )}
              </Typography>
            </Box>
            <Box mb={2}>
              <Typography variant="subtitle2">
                {t('Select types of reports to be exported')}:
              </Typography>
            </Box>
            <Table>
              <TableHead>
                <TableRow>
                  <TableCell padding="checkbox">
                    <Checkbox
                      color="primary"
                      indeterminate={numSelected > 0 && numSelected < rowCount}
                      checked={rowCount > 0 && numSelected === rowCount}
                      onChange={(event) => onSelectAllClick(event, data)}
                      inputProps={{ 'aria-label': 'select all' }}
                    />
                  </TableCell>
                  <TableCell align="left">{t('Report Type')}</TableCell>
                </TableRow>
              </TableHead>
              <TableBody>{renderRows()}</TableBody>
            </Table>
            <Box p={3} m={4} bgcolor="#F5F5F5">
              <Typography variant="subtitle2">
                {t(
                  'Upon clicking export button, report will be generated and send to your email address when ready.',
                )}
              </Typography>
            </Box>
          </DialogContainer>
        </DialogContent>
        <DialogFooter>
          <DialogActions>
            <Button
              onClick={() => {
                setDialogOpen(false);
                setSelected([]);
              }}
            >
              {t('CANCEL')}
            </Button>
            <Button
              type="submit"
              color="primary"
              variant="contained"
              onClick={submitFormHandler}
              data-cy="button-submit"
            >
              {t('Export')}
            </Button>
          </DialogActions>
        </DialogFooter>
      </Dialog>
    </>
  );
}
