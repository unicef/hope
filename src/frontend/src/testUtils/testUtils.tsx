import { ReactElement } from 'react';
import { BrowserRouter } from 'react-router-dom';
import { render, RenderOptions } from '@testing-library/react';
import { Formik } from 'formik';
import noop from 'lodash/noop';
import { TestProviders } from './testProviders';
import { ProgramStatus } from '@generated/graphql';
import { MockLink } from '@apollo/client/testing';

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) =>
  render(
    <>
      <BrowserRouter>
        <Formik initialValues={{}} onSubmit={noop}>
          {ui}
        </Formik>
      </BrowserRouter>
    </>,
    {
      wrapper: TestProviders,
      ...options,
    },
  );

export * from '@testing-library/react';
export { customRender as render };

let m_w = 123456789;
let m_z = 987654321;
const mask = 0xffffffff;

// Takes any integer
export const seed = (i) => {
  m_w = (123456789 + i) & mask;
  m_z = (987654321 - i) & mask;
};

// Returns number between 0 (inclusive) and 1.0 (exclusive),
// just like Math.random().
export const random = () => {
  m_z = (36969 * (m_z & 65535) + (m_z >> 16)) & mask;
  m_w = (18000 * (m_w & 65535) + (m_w >> 16)) & mask;
  let result = ((m_z << 16) + (m_w & 65535)) >>> 0;
  result /= 4294967296;
  return result;
};

export class ApolloLoadingLink extends MockLink {
  constructor() {
    super([]);
  }

  request() {
    return null;
  }
}

export const fakeContextProgram = {
  selectedProgram: {
    id: '1',
    name: 'someName',
    status: ProgramStatus.Active,
    pduFields: null,
    dataCollectingType: {
      id: '1',
      householdFiltersAvailable: true,
      individualFiltersAvailable: true,
      label: 'data collecting type',
      code: '123',
      type: 'full',
      children: null,
    },
  },
  setSelectedProgram: () => {},
  isActiveProgram: true,
  isSocialDctType: false,
  isStandardDctType: true,
  programHasPdu: false,
};
