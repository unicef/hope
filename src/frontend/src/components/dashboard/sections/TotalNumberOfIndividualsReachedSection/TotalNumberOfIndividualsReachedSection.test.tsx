import * as React from 'react';
import { render } from '../../../../testUtils/testUtils';
import { TotalNumberOfIndividualsReachedSection } from './TotalNumberOfIndividualsReachedSection';
import { fakeSectionIndividualsReached } from '../../../../../fixtures/dashboard/fakeSectionIndividualsReached';
import { fakeChartIndividualsReachedByAgeAndGender } from '../../../../../fixtures/dashboard/fakeChartIndividualsReachedByAgeAndGender';
import { fakeChartIndividualsWithDisabilityReachedByAge } from '../../../../../fixtures/dashboard/fakeChartIndividualsWithDisabilityReachedByAge';

describe('components/dashboard/TotalNumberOfIndividualsReachedSection', () => {
  it('should render', () => {
    const { container } = render(
      <TotalNumberOfIndividualsReachedSection
        data={fakeSectionIndividualsReached}
        chartDataIndividuals={fakeChartIndividualsReachedByAgeAndGender}
        chartDataIndividualsDisability={
          fakeChartIndividualsWithDisabilityReachedByAge
        }
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
