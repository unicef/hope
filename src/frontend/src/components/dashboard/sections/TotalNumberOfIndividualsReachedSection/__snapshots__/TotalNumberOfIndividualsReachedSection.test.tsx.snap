// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/dashboard/TotalNumberOfIndividualsReachedSection should render 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 sc-kLhKbu bmwOMh css-1ps6pg7-MuiPaper-root"
    color="#345DA0"
  >
    <div
      class="sc-egkSDF DXwdF"
    >
      TOTAL NUMBER OF INDIVIDUALS REACHED
    </div>
    <div
      class="MuiGrid-root MuiGrid-container css-9cyib4-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        data-cy="total-number-of-individuals-reached"
      >
        <div
          class="sc-khLCKb hScFTY"
        >
          0
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <div
          bg="#D9E2EF"
          class="sc-ghWlax gTXmEW"
          color="#023F90"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
            data-testid="PersonIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4m0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
